Fix InlineArray and some other corner cases in Swift lowering

jkoritzinsky opened the pull request:
Fix the case mentioned in https://github.com/dotnet/runtime/pull/99294#issuecomment-1978900870

Fixes InlineArray support

jkoritzinsky committed:
Use TypeWithRepeatedFields in the FieldLayoutIntervalCalculator to ensure we have correct inlinearray support and fix unaligned lowering calculations

jakobbotsch commented on src/coreclr/tools/Common/JitInterface/SwiftPhysicalLowering.cs:
```diff
@@ -154,21 +154,21 @@ public List<CorInfoType> GetLoweredTypeSequence()
                         }
 
-                        if (remainingIntervalSize == 1)
+                        while (remainingIntervalSize > 0)
```
This loop is missing a `remainingIntervalSize--` it looks like.

jkoritzinsky committed:
Add missing decrement

jkoritzinsky committed:
Add stress tests for the lowering and update the algorithm to handle it.

The following algorithm changes were made:

- Don't merge an Opaque interval and and Empty interval during construction.
- Introduce a post-processing phase that removes Empty intervals and merges Opaque intervals
  that used to have Empty intervals between them, as long as the opaque intervals are within the
  same pointer-size block.
- Use the opaque interval's offset and size when breaking it up.
- Allow the opaque splitting to split up higher sizes if the lower sizes have improved the alignment.

jkoritzinsky committed:
Fix ILVerify build and simplify now that these concepts are in the base type system.

jakobbotsch commented:
Do we want to check in all the tests, or should we reduce them to a smaller set once everything is passing (perhaps making sure we include the specific cases that were investigated)? I wasn't planning on checking all my tests in in #99294 since it seems like a lot of source code to commit.

jkoritzinsky committed:
Extend the CORINFO_SWIFT_LOWERING type to specify offsets.

Update tests that need to specify non-natural offsets to specify them.
Reduce Swift test suite to a representative sample of the various cases that have captured failures.
